components/core/src/clp/ffi/KeyValuePairLogEvent.cpp

```diff
@@ -20,9 +23,87 @@
 using clp::ir::EightByteEncodedTextAst;
 using clp::ir::FourByteEncodedTextAst;
 using std::string;
+using std::vector;
 
 namespace clp::ffi {
 namespace {
+/**
+ * Callback to handle JSON exceptions. When we cannot throw a captured JSON exception (i.e., in a
+ * destructor), there must be a way to not crash the program but also let the caller be aware of the
+ * exception. This concept plays the role to allow users to customize a callback handler when
+ * catching an exception.
+ */
+template <typename Func>
+concept JsonExceptionCallbackConcept = std::is_invocable_v<Func, nlohmann::json::exception const&>;
+
+/**
+ * Class for iterating the schema tree using DFS. It contains a JSON map object representing itself,
+ * which will be constructed by traversing through its children. When all children has been visited,
+ * the iterator will be popped out from the DFS stack and emplace the constructed map to its parent
+ * as a subtree.
+ */
+template <JsonExceptionCallbackConcept JsonExceptionCallback>
+class SchemaTreeDfsIterator {
```

Comment:
This class is fairly confusing on a first read, and only really makes sense in the context of the serialization code that uses it.

I think our code guidelines say that classes are supposed to be at the top, but I would maybe consider moving this to right before its use.

I would also rewrite this description to something like

```
Helper class for iterating over a sub-tree of the schema tree in a depth-first manner. 
KeyValuePairLogEvent::serialize_to_json uses a stack of instances of this class to perform depth-first serialization of JSON objects.
Each instance of this class holds state from the perspective of a single node in the schema tree.
This includes a JSON map representing itself, which gets populated while traversing over its children.
Once all its children have been visited, the iterator instance gets popped off the DFS stack, and emplaces the contents of its JSON map into its parent as a subtree.
```

Reply:
The description looks good to me. For moving the definition around, what do you think? @kirkrodrigues